File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys
input = sys.stdin.readline

n, c = map(int, input().split())
house = [int(input()) for _ in range(n)]
house.sort()

start = 1
end = house[-1]
result = 0

while start <= end:
    flag = house[0]
    mid = (start + end) // 2
    cnt = 1

    for i in range(1, n):
        if house[i] - flag >= mid:
            cnt += 1
            flag = house[i]
    
    if cnt >= c:
        result = mid
        start = mid + 1
    else:
        end = mid - 1

print(result)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    private static int N, C;
    private static int[] house;
    
    private static int calcMinDist(int interval) {
        int prev = house[0];
        int wifi = 1;
        
        for (int i = 1; i < N; i++) {
        	if (house[i] - prev >= interval) {
        		prev = house[i];
        		wifi++;
        		
        		if (wifi == C) {
        			break;
        		}
        	}
        }
        
        return wifi;
    }
    
    private static int binarySearch() {
        int s = 0;
        int e = house[N-1];
        int res = 0;
        
        while (s <= e) {
            int mid = (s + e) / 2;
            int temp = calcMinDist(mid);
            
            if (temp < C) {
            	e = mid - 1;
            } else {
            	res = Math.max(res, mid);
            	s = mid + 1;
            }
        }
        
        return res;
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer nums = new StringTokenizer(br.readLine());
        N = Integer.parseInt(nums.nextToken());
        C = Integer.parseInt(nums.nextToken());
        house = new int[N];
        
        for (int i = 0; i < N; i++) {
            int h = Integer.parseInt(br.readLine());
            house[i] = h;
        }
        
        Arrays.sort(house);
        System.out.println(binarySearch());
    }

}